File: 프로그래머스/2/42626. 더 맵게/더 맵게.cpp
#include <string>
#include <vector>
#include <queue>
#include <functional>
#include <iostream>

using namespace std;

int solution(vector<int> scoville, int K) {
    int answer = 0;
    priority_queue<int, vector<int>, greater<int>> pq;
    for(int s: scoville){
        pq.push(s);
    }
    while (pq.size() > 1 && pq.top() < K){
        int s_first = pq.top();
        pq.pop();
        int s_second = pq.top();
        pq.pop();
        
        pq.push(s_first + s_second * 2);
        answer ++;
    }
    if (pq.size() == 1 && pq.top() < K){
        return -1;
    }
    return answer;
}

File: 프로그래머스/2/42626. 더 맵게/더 맵게.java
import java.util.*;
import java.util.stream.Collectors;


class Solution {
    public int solution(int[] scoville, int K) {
        int answer = 0;
        
        PriorityQueue<Integer> pq = Arrays.stream(scoville)
            .boxed()
            .collect(Collectors.toCollection(PriorityQueue::new));
        
        while (pq.size() > 1 && pq.peek() < K){
            int f1 = pq.poll();
            int f2 = pq.poll();

            pq.add(f1 + f2 * 2);
            answer ++;
            
        }
        
        if (pq.peek() < K) return -1;
        else return answer;
    }
}